### ISSUE TITLE
eq3loop: eq3loop_write_master() mmd_hmip: not enought space in the buffers

### ISSUE THREAD

**Issue by jens-maus** [jens-maus/RaspberryMatic#703]
**Describe the bug**
Under some rare race conditions it seems that the eq3loop kernel driver runs into some issues eventually blocking `/dev/mmd_hmip` or `/dev/mmd_bidcos` resulting in some crashes of, e.g. `HMIPServer` and `multimac` complaining about "Bad address".

**To Reproduce**

Look into your `/var/log/messages` log file for `not enough space in the buffers` kernel messages from eq3loop which should look like:

```
Sep  3 21:24:57 ccupi user.warn multimac: UpstreamCharConnection slave device mmd_hmip write error: Bad address
Sep  3 21:24:57 ccupi user.err kernel: [341836.403016] eq3loop: eq3loop_write_master() mmd_hmip: not enought space in the buffers. free
 space = 23, required space = 48
Sep  3 21:24:59 ccupi user.warn multimac: UpstreamCharConnection slave device mmd_hmip write error: Bad address
Sep  3 21:24:59 ccupi user.info kernel: [341836.403024] eq3loop: eq3loop_write_master() retrun error:
Sep  3 21:24:59 ccupi user.err kernel: [341838.323322] eq3loop: eq3loop_write_master() mmd_hmip: not enought space in the buffers. free space = 23, required space = 28
Sep  3 21:28:57 ccupi user.warn multimac: UpstreamCharConnection slave device mmd_hmip write error: Bad address
Sep  3 21:28:57 ccupi user.info kernel: [341838.323330] eq3loop: eq3loop_write_master() retrun error:
Sep  3 21:28:57 ccupi user.err kernel: [342076.404956] eq3loop: eq3loop_write_master() mmd_hmip: not enought space in the buffers. free space = 39, required space = 48
Sep  3 21:29:00 ccupi user.warn multimac: UpstreamCharConnection slave device mmd_hmip write error: Bad address
Sep  3 21:29:00 ccupi user.info kernel: [342076.404963] eq3loop: eq3loop_write_master() retrun error:
Sep  3 21:29:00 ccupi user.err kernel: [342078.773497] eq3loop: eq3loop_write_master() mmd_hmip: not enought space in the buffers. free space = 39, required space = 42
Sep  3 21:29:26 ccupi user.warn multimac: UpstreamCharConnection slave device mmd_hmip write error: Bad address
Sep  3 21:29:26 ccupi user.info kernel: [342078.773503] eq3loop: eq3loop_write_master() retrun error:
Sep  3 21:29:26 ccupi user.err kernel: [342104.618264] eq3loop: eq3loop_write_master() mmd_hmip: not enought space in the buffers. free space = 1, required space = 48
Sep  3 21:29:27 ccupi user.warn multimac: UpstreamCharConnection slave device mmd_hmip write error: Bad address
Sep  3 21:29:27 ccupi user.info kernel: [342104.618271] eq3loop: eq3loop_write_master() retrun error:
Sep  3 21:29:27 ccupi user.err kernel: [342105.593457] eq3loop: eq3loop_write_master() mmd_hmip: not enought space in the buffers. free space = 1, required space = 32
Sep  3 21:29:27 ccupi user.warn multimac: UpstreamCharConnection slave device mmd_hmip write error: Bad address
Sep  3 21:29:27 ccupi user.info kernel: [342105.593465] eq3loop: eq3loop_write_master() retrun error:
Sep  3 21:29:27 ccupi user.err kernel: [342106.134816] eq3loop: eq3loop_write_master() mmd_hmip: not enought space in the buffers. free space = 1, required space = 32
```

**Expected behavior**
Under optimal conditions these messages should NEVER appear and eq3loop should have enough circular buffers to store all events at a time. But obviously this does not seem to be the case somehow.

**System information (please complete the following information):**
 - Version: RaspberryMatic 3.47.15.20190831
 - Hardware: TinkerboardS

**Additional context**
When the above syslog output appears, the following output can be found in the `/var/log/hmserver.log`:

```
Sep 3 21:18:40 de.eq3.cbcs.server.core.otau.DeviceBackgroundUpdateSubsystem INFO  [vert.x-eventloop-thread-2] SYSTEM: Eventlistener Han
dler utilization: 0/50 used 
Sep 3 21:55:42 de.eq3.cbcs.lib.nrjavaserialdevice.internal.SerialFrameDecoder ERROR [Thread-6] Last frame was not complete. Aborting decoding of that frame. 
Sep 3 22:05:17 de.eq3.lib.util.dynamics.GenericFactory INFO  [main] @GenericFactory 
Sep 3 22:05:17 de.eq3.lib.util.dynamics.GenericFactory INFO  [main] created instance of HMServerConfiguration with parameter(s) 
```

Note the 30 seconds delay between the last INFO message and the next ERROR message which might be the point where HMServer was crashing unexpected. Furthermore, then Monit identifies HMIPServer as not working properly and restarting it (see 22:05:17 INFO messages) and the following Monit output:

```
Sep  3 22:03:37 ccupi user.warn monit[1116]: 'HMIPServer' failed protocol test [DEFAULT] at [localhost]:39292 [TCP/IP] -- Connection timed out
Sep  3 22:03:59 ccupi user.warn monit[1116]: 'HMIPServer' failed protocol test [DEFAULT] at [localhost]:39292 [TCP/IP] -- Connection timed out
Sep  3 22:04:22 ccupi user.warn monit[1116]: 'HMIPServer' failed protocol test [DEFAULT] at [localhost]:39292 [TCP/IP] -- Connection timed out
Sep  3 22:04:45 ccupi user.warn monit[1116]: 'HMIPServer' failed protocol test [DEFAULT] at [localhost]:39292 [TCP/IP] -- Connection timed out
Sep  3 22:05:08 ccupi user.err monit[1116]: 'HMIPServer' failed protocol test [DEFAULT] at [localhost]:39292 [TCP/IP] -- Connection timed out
Sep  3 22:05:08 ccupi user.info monit[1116]: 'HMIPServer' trying to restart
Sep  3 22:05:08 ccupi user.info monit[1116]: 'HMIPServer' restart: '/etc/init.d/S62HMServer restart'
```

All this seems to point at some issue (root cause still unknown) which seem to start in eq3loop not being able to fill all buffers and thus resulting in /dev/mmd_hmip not working properly and thus HMIPServer seem to crash. Perhaps increasing the buffer size could solve this problem?

**Comment by alexreinert** [comment]
I think, that the buffer size of 1024 is big enough: The messages provided by the radio module are smaller than 50 bytes, so at least 20 messages should fit into the buffer. Sending 1024 bytes to the radio module should be sent to the radio module within 10ms (communication is done with 115000 baud). That means, that there must be more than 2 messages per millisecond.
Combined with the interrupt 0x0c issue (which I have never seen), I think there is an issue outside the raw-uart and the eq3_char_loop. My guess would be some interrupt or locking issues within the kernel itself, resulting in a slow handling of the interrupt handlers of the kernel modules. How many soft interrupts do you see in the system? I remember an issue, the the usb otg kernel module of the Tinkerboard produced a huge amount of soft interrupts when more than one usb devices were attached. Depending on the sheduler this issue blocked one core.

**Comment by alexreinert** [comment]
Additionaly: Increasing the buffer may resolve the error messages, but the this could result in a delay of the message processing that is higher than the required 5ms between the first sent byte of the request and the last byte of the response.

**Comment by jens-maus** [comment]
I also agree and have the same feeling that just increasing BUFSIZE might not solce the root cause here. But I currently have no real testcase to reproduce the issue and that issue seem to happen only very rarely (I just have it every 2-3 months). So the question is really how to further debug this case and add some code just in case it happens again in future. Thus, increasing BUFSIZE was done in the hope to somewhat workaround the issue because as I outlined above it seems to be related to crashes of HmIPServer which seems to be more severe to me than having some frames exceeding the 5ms range.

**Comment by jens-maus** [comment]
BTW: What do you think @alexreinert? Shouldn't it be also possible to change the following while()?

https://github.com/jens-maus/RaspberryMatic/blob/a00e3c4c23ca30d96b17134e4bd63b4585ae4a7c/buildroot-external/package/eq3_char_loop/eq3_char_loop.c#L347

IMHO it should be possible to simply loop there until there is enough space in the circular buffer for putting `count` items in there. Currently, the while loop exists as soon as there is some space (>0).

**Comment by alexreinert** [comment]
On a first sight, I would say, that this is a bad idea. This construct prevents the write operation from an endless loop.

**Comment by jens-maus** [comment]
yes, that can be the case. But under which other circumstances can the buffer then be too small? How can this happen? As you know the whole serial communication stuff way better than me I hope we can find some better explanation than simply BUFSIZE being to small in busy cases (which I currently simply have to assume, thus increasing BUFSIZE as a workaround to keep things going somehow).

**Comment by alexreinert** [comment]
As written, I do not think, that the buffer size is to small, it should not happen, that there are more than 20 messages in 10 msec.
And as said, my guess is, that there are too many other interrupt handlers, which block the raw-uart and the eq3_char_loop.
Some questions: Did the interrupt 0x0c issue and this issue occur in a temporal matter? Are you using the irq balance daemon? How many usb devices are connected?

**Comment by jens-maus** [comment]
No, these two things (unhandled 0xc interrupt and eq3loop error) seem to be not related to each other. At least I cannot spot these two things at the same time here in the syslog/dmesg output. And yes, I am using the irq balance daemon per default and I have no USB devices connected to the Tinkerboard. Thus, these are the only USB devices listed via `lsusb`:

```
Bus 002 Device 002: ID 0bda:481a
Bus 001 Device 001: ID 1d6b:0002
Bus 001 Device 002: ID 05e3:0610
Bus 002 Device 001: ID 1d6b:0002
```

**Comment by jens-maus** [comment]
At least I cannot see wild things going on here. As pointed out above this is the output right before the first "not enough space in the buffers" messages happens:

```
[...]
Sep  3 21:19:40 ccupi user.warn rfd: XmlRpc transport failed (first try), retrying...
Sep  3 21:19:50 ccupi user.err rfd: XmlRpcClient error calling event({[methodName:"event",params:{"BidCos-RF_java","LEQ1064334:4","CONTROL_MODE",0}],[methodName:"event",params:{"BidCos-RF_java","LEQ1064334:4","FAULT_REPORTING",0}],[methodName:"event",params:{"BidCos-RF_java","LEQ1064334:4","BATTERY_STATE",3.100000}],[methodName:"event",params:{"BidCos-RF_java","LEQ1064334:4","VALVE_STATE",0}],[methodName:"event",params:{"BidCos-RF_java","LEQ1064334:4","BOOST_STATE",0}],[methodName:"event",params:{"BidCos-RF_java","LEQ1064334:4","ACTUAL_TEMPERATURE",23.500000}],[methodName:"event",params:{"BidCos-RF_java","LEQ1064334:4","SET_TEMPERATURE",12.000000}],[methodName:"event",params:{"BidCos-RF_java","LEQ1064334:4","PARTY_TEMPERATURE",5.000000}],[methodName:"event",params:{"BidCos-RF_java","LEQ1064334:4","PARTY_START_TIME",0}],[methodName:"event",params:{"BidCos-RF_java","LEQ1064334:4","PARTY_START_DAY",1}],[methodName:"event",params:{"BidCos-RF_java","LEQ1064334:4","PARTY_START_MONTH",1}],[methodName:"event",params:{"BidCos-R
Sep  3 21:19:50 ccupi user.err rfd: XmlRpc transport error
Sep  3 21:24:57 ccupi user.warn multimac: UpstreamCharConnection slave device mmd_hmip write error: Bad address
Sep  3 21:24:57 ccupi user.err kernel: [341836.403016] eq3loop: eq3loop_write_master() mmd_hmip: not enought space in the buffers. free space = 23, required space = 48
[...]
```

So there are some XmlRpc transport error in the syslog before it happened, but between the last XmlRpc error and the first "not enought space" messages there are several minutes in between and then first multimac outputs some warning about not being able to access /dev/mmd_hmip and then the problems starts to emerge until HMIPServer crashed as it seems.

```
[...]
Sep 3 21:18:29 de.eq3.cbcs.server.core.otau.DeviceBackgroundUpdateSubsystem INFO  [vert.x-eventloop-thread-2] SYSTEM: 0 Accesspoints in
 Queue 
Sep 3 21:18:32 de.eq3.cbcs.legacy.bidcos.rpc.LegacyServiceHandler INFO  [vert.x-worker-thread-4] init finished 
Sep 3 21:18:30 de.eq3.cbcs.lib.remotecommadapter.device.service.AccessPointWatchdog INFO  [vert.x-eventloop-thread-0] SYSTEM: Access Po
int Watchdog running periodic connection check... 
Sep 3 21:18:32 de.eq3.cbcs.server.core.otau.DeviceBackgroundUpdateSubsystem INFO  [vert.x-eventloop-thread-2] SYSTEM: Permanent-/Burstl
istener Handler utilization: 0/50 used 
Sep 3 21:18:40 de.eq3.cbcs.server.core.otau.DeviceBackgroundUpdateSubsystem INFO  [vert.x-eventloop-thread-2] SYSTEM: Eventlistener Han
dler utilization: 0/50 used 
Sep 3 21:55:42 de.eq3.cbcs.lib.nrjavaserialdevice.internal.SerialFrameDecoder ERROR [Thread-6] Last frame was not complete. Aborting de
coding of that frame. 
Sep 3 22:05:17 de.eq3.lib.util.dynamics.GenericFactory INFO  [main] @GenericFactory 
Sep 3 22:05:17 de.eq3.lib.util.dynamics.GenericFactory INFO  [main] created instance of HMServerConfiguration with parameter(s) 
Sep 3 22:05:17 de.eq3.lib.util.dynamics.GenericFactory INFO  [main] passed 1 parameter(s), in declarative order [String] 
Sep 3 22:05:22 de.eq3.cbcs.vertx.management.VertxManager INFO  [Thread-0] SYSTEM: added for deployment [HMIPTRXWriterWorker] (1) *worker
[...]
```

As you can hopefully see from the above HM-Server.log output, everything seem to worked fine until 21:18:40 and then there was a large delay until the next output of the `de.eq3.cbcs.lib.nrjavaserialdevice.internal.SerialFrameDecoder` @ 21:55:42 and then monitor actually restarted HMIPServer @ 22:05:17 after having identified that it was not working properly anymore on port 39292.

**Comment by alexreinert** [comment]
* The problem also occurs on the CCU2 --> This is likely no issue regarding unreflected registers in on a multicore cpu
* The Access Point Watchdog running periodic connection check normally runs every 5 minutes. Last time it run in this case was 21:18:30, the run at 21:23:30 is missing --> something blocks the Watchdog runs inside the HMIPServer
* Only the buffer in the direction multimacd to HMIPserver is affected. RFD should get the same data using /dev/mmd_bidcos but has no issue. --> It is likely no message overflow
* Restarting the HMIPServer solves the problem

My conclusion: There is a race condition inside the HMIPserver or inside the eq3_char_loop which prevents reading the data from the buffer of the /dev/mmd_hmip slave device. Increasing the buffer would only increase the time until the crash occurs.

Maybe you can reach, that I get the source part of the HMIPServer there the nrjavaserial lib is called. Maybe I could see some issue there.

**Comment by alexreinert** [comment]
Issue closed? Did eQ-3 fix it inside the HmIOServer?

**Comment by asman-hm** [comment]
I am not sure, but I think this issue came back. I installed release 3.49.17.20200131 yesterday at about 1 p.m. (was on release 3.47.22.20191130 before) and my system crashed tonight with the following errors (log is full with hundreds of these lines):

2020-02-04,03:33:49,Information,192.168.0.60,user,kernel,[53731.225087] eq3loop: eq3loop_write_master() return error: -14
2020-02-04,03:33:47,Error,192.168.0.60,user,kernel,"[53731.225081] eq3loop: eq3loop_write_master() mmd_hmip: not enough space in buffers. free space = 5, required space = 37"
2020-02-04,03:33:47,Information,192.168.0.60,user,kernel,[53727.863589] eq3loop: eq3loop_write_master() return error: -14
2020-02-04,03:33:44,Error,192.168.0.60,user,kernel,"[53727.863583] eq3loop: eq3loop_write_master() mmd_hmip: not enough space in buffers. free space = 5, required space = 37"


Nothing was working any more, only a reboot this morning helped. For the moment I will stay on this version. If you need more information, feel free to contact me.


TinkerboardS
CCU-Historian
Cup-Deamon
approx. 20+ HMIP-devices and 80+ HM-devices

**Comment by pannal** [comment]
I'm experiencing the issue as well on a CCU3 with RaspberryMatic-3.59.6.20210703.

```
[Tue Jul 27 16:05:01 2021] eq3loop: eq3loop_write_master() return error: -14
[Tue Jul 27 16:06:02 2021] eq3loop: eq3loop_write_master() mmd_hmip: not enough space in buffers. free space = 4, required space = 34
[Tue Jul 27 16:06:02 2021] eq3loop: eq3loop_write_master() return error: -14
[Tue Jul 27 16:06:23 2021] eq3loop: eq3loop_write_master() mmd_hmip: not enough space in buffers. free space = 4, required space = 49
[Tue Jul 27 16:06:23 2021] eq3loop: eq3loop_write_master() return error: -14
[Tue Jul 27 16:06:53 2021] eq3loop: eq3loop_write_master() mmd_hmip: not enough space in buffers. free space = 4, required space = 35
[Tue Jul 27 16:06:53 2021] eq3loop: eq3loop_write_master() return error: -14
[Tue Jul 27 16:08:50 2021] eq3loop: eq3loop_write_master() mmd_hmip: not enough space in buffers. free space = 4, required space = 34
[Tue Jul 27 16:08:50 2021] eq3loop: eq3loop_write_master() return error: -14
[Tue Jul 27 16:09:50 2021] eq3loop: eq3loop_write_master() mmd_hmip: not enough space in buffers. free space = 4, required space = 34
[Tue Jul 27 16:09:50 2021] eq3loop: eq3loop_write_master() return error: -14
[Tue Jul 27 16:10:34 2021] eq3loop: eq3loop_write_master() mmd_hmip: not enough space in buffers. free space = 4, required space = 47
[Tue Jul 27 16:10:34 2021] eq3loop: eq3loop_write_master() return error: -14
[Tue Jul 27 16:11:24 2021] eq3loop: eq3loop_write_master() mmd_hmip: not enough space in buffers. free space = 4, required space = 34
[Tue Jul 27 16:11:24 2021] eq3loop: eq3loop_write_master() return error: -14
[Tue Jul 27 16:12:06 2021] eq3loop: eq3loop_write_master() mmd_hmip: not enough space in buffers. free space = 4, required space = 47
[Tue Jul 27 16:12:06 2021] eq3loop: eq3loop_write_master() return error: -14
[Tue Jul 27 16:12:21 2021] eq3loop: eq3loop_write_master() mmd_hmip: not enough space in buffers. free space = 4, required space = 49
[Tue Jul 27 16:12:21 2021] eq3loop: eq3loop_write_master() return error: -14
[Tue Jul 27 16:12:33 2021] eq3loop: eq3loop_write_master() mmd_hmip: not enough space in buffers. free space = 4, required space = 34
[Tue Jul 27 16:12:33 2021] eq3loop: eq3loop_write_master() return error: -14
[Tue Jul 27 16:13:43 2021] eq3loop: eq3loop_write_master() mmd_hmip: not enough space in buffers. free space = 4, required space = 34
```

I've got about 6 HMIP-WTH/STH, two HMIP-ETRV and a HMIP-FALMOT-C12 connected.

**Comment by mark77-DE** [comment]
Good morning,

I had the same issues last night/this morning, first I thought it is just a Firmware problem.
After updating the main unit the error still reoccurred, so I also updated my two RaspberryMatic LAN Gateways.
Happily after that, no errors showed up.
Last night, because of a battery low warning, I replaced the battery of an HM-PBI-4-FM, which was empty again this morning.
After replacing it to a second new battery, trouble started all over again..

So for me, all this was caused by an defective HM-PBI-4-FM.
